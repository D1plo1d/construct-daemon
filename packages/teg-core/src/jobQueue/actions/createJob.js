export const CREATE_JOB = 'teg/jobQueue/CREATE_JOB'

const createJob = ({ job, jobFiles }) => ({
  type: CREATE_JOB,
  payload: {
    job,
    jobFiles,
  },
})

export default createJob
