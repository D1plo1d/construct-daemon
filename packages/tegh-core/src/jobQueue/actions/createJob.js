import tmp from 'tmp-promise'

import fs from '../../util/promisifiedFS'
import Job from '../types/Job'
import JobFile from '../types/JobFile'

export const CREATE_JOB = 'tegh/jobQueue/CREATE_JOB'

const createJob = ({ files, name }) => async (dispatch) => {
  if (name == null) {
    throw new Error('name cannot be null')
  }

  if (files == null) {
    throw new Error('files cannot be null')
  }

  const job = Job({ name })
  const jobFiles = {}

  // eslint-disable-next-line no-restricted-syntax
  for (const file of files) {
    if (typeof file.name !== 'string') {
      throw new Error('file name must be a string')
    }

    if (typeof file.content !== 'string') {
      throw new Error('file content must be a string')
    }

    const tmpFile = await tmp.file()
    const filePath = tmpFile.path
    await fs.writeFileAsync(filePath, file.content)

    const jobFile = JobFile({
      jobID: job.id,
      name: file.name,
      filePath,
      isTmpFile: true,
      quantity: 1,
    })

    jobFiles[jobFile.id] = jobFile
  }

  const action = {
    type: CREATE_JOB,
    payload: {
      job,
      jobFiles,
    },
  }

  return dispatch(action)
}

export default createJob
