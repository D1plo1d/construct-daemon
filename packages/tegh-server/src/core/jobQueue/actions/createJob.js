import path from 'path'
import tmp from 'tmp-promise'
import untildify from 'untildify'
import { List } from 'immutable'

import fs from '../../util/promisifiedFS'
import validateCommandsFileExtension from '../../util/validateCommandsFileExtension'
import Job from '../types/Job'
import JobFile from '../types/JobFile'

const normalize = filePath => path.normalize(untildify(filePath))

export const CREATE_JOB = 'tegh/jobQueue/CREATE_JOB'

const createJob = ({ files, name }) => {
  return async (dispatch) => {

    if (name == null) {
      throw new Error('name cannot be null')
    }

    if (files == null) {
      throw new Error('files cannot be null')
    }

    const job = Job({ name })
    const jobFiles = {}

    for (const file of files) {
      if (typeof file.name !== 'string') {
        throw new Error('file name must be a string')
      }

      if (typeof file.content !== 'string') {
        throw new Error('file content must be a string')
      }

      const tmpFile = await tmp.file()
      const filePath = tmpFile.path
      await fs.writeFileAsync(filePath, file.content)

      const jobFile = JobFile({
        jobID: job.id,
        name: file.name,
        filePath,
        isTmpFile: true,
        quantity: 1,
      })

      jobFiles[jobFile.id] = jobFile
    }

    const action = {
      type: CREATE_JOB,
      payload: {
        job,
        jobFiles,
      },
    }

    return dispatch(action)
  }
}

export default createJob
