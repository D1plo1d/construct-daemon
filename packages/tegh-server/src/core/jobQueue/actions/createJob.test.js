import tmp from 'tmp-promise'

import fs from '../../util/promisifiedFS'
import expectToMatchImmutableSnapshot from '../../util/testing/expectToMatchImmutableSnapshot'
import createJob from './createJob'

describe('createJob', () => {
  it('creates a CREATE_JOB action', async function() {
    const name = 'test_test_test'
    const files = [
      {
        name: 'file_A',
        content: `G28`,
      },
      {
        name: 'file_B',
        content: `G1 X10\nG1 Y10\nG1 Z10`
      }
    ]

    const dispatch = action => action

    let result = await createJob({
      files,
      name
    })(dispatch)

    expectToMatchImmutableSnapshot({
      result,
      redactions: [
        ['payload', 'job', 'id'],
        ['payload', 'job', 'createdAt'],
        ['payload', 'jobFiles'],
      ],
    })

    const jobFiles = Object.values(result.payload.jobFiles)
    for (const jobFile of jobFiles) {
      expectToMatchImmutableSnapshot({
        result: jobFile,
        redactions: [
          ['id'],
          ['filePath'],
          ['jobID'],
        ],
      })
      const content = await fs.readFileAsync(jobFile.filePath, 'utf8')
      expect(content).toMatchSnapshot()

      /* clean up tmp file */
      await fs.unlinkAsync(jobFile.filePath)
    }

  })
})
