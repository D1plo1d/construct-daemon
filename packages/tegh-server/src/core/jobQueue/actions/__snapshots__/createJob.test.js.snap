// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJob creates a CREATE_JOB action 1`] = `
Object {
  "payload": Object {
    "job": Object {
      "createdAt": "[REDACTED]",
      "id": "[REDACTED]",
      "name": "test_test_test",
    },
    "jobFiles": "[REDACTED]",
  },
  "type": "tegh/jobQueue/CREATE_JOB",
}
`;

exports[`createJob creates a CREATE_JOB action 2`] = `
Object {
  "filePath": "[REDACTED]",
  "id": "[REDACTED]",
  "isTempFile": true,
  "name": "file_A",
  "quantity": 1,
}
`;

exports[`createJob creates a CREATE_JOB action 3`] = `"G28"`;

exports[`createJob creates a CREATE_JOB action 4`] = `
Object {
  "filePath": "[REDACTED]",
  "id": "[REDACTED]",
  "isTempFile": true,
  "name": "file_B",
  "quantity": 1,
}
`;

exports[`createJob creates a CREATE_JOB action 5`] = `
"G1 X10
G1 Y10
G1 Z10"
`;
